### DISCUSSION TITLE
c++ sdk ubuntu build erro

### DISCUSSION THREAD

**Discussion by NoError404** [milvus-io/milvus#4892] [Q&A and General discussion]
this is cmakeError.log

[CMakeError.log](https://github.com/milvus-io/milvus/files/6241561/CMakeError.log)

**Comment by fishpenguin** [comment]
>src.c:(.text+0x46): undefined reference to `pthread_create'
/usr/bin/ld: src.c:(.text+0x52): undefined reference to `pthread_detach'
/usr/bin/ld: src.c:(.text+0x63): undefined reference to `pthread_join'

please refer [this](https://stackoverflow.com/questions/1662909/undefined-reference-to-pthread-create-in-linux).

**Reply by NoError404** [reply]
**thanks for u reply.I have added set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread") in cmakelists.txt ,  but it did't work.i wonder if the problem caused by what the terminal show?**

[ 16%] Built target zlib_ep
[ 18%] Performing build step for 'grpc_ep'
make[3]: warning: -j8 forced in submake: resetting jobserver mode.
[MAKE]    Generating cache.mk
[AR]      Creating /home/hustzz/下载/milvus_c++/sdk/cmake_build/grpc_ep-prefix/src/grpc_ep/libs/opt/libaddress_sorting.a
[AR]      Creating /home/hustzz/下载/milvus_c++/sdk/cmake_build/grpc_ep-prefix/src/grpc_ep/libs/opt/libares.a

make[3]: *** [Makefile:993：stop] 错误 1
make[3]: *** 正在等待未完成的任务....
DEPENDENCY ERROR

The target you are trying to run requires an OpenSSL implementation.
Your system doesn't have one, and either the third_party directory
doesn't have it, or your compiler can't build BoringSSL.

Please consult INSTALL to get more information.

If you need information about why these tests failed, run:

  make run_dep_checks

make[2]: *** [CMakeFiles/grpc_ep.dir/build.make:115：grpc_ep-prefix/src/grpc_ep-stamp/grpc_ep-build] 错误 2
make[1]: *** [CMakeFiles/Makefile2:285：CMakeFiles/grpc_ep.dir/all] 错误 2
make: *** [Makefile:130：all] 错误 2

**Reply by yhmo** [reply]
A workaround:

- import the https://github.com/madler/zlib to Gitee, you can refer this article: https://www.pianshen.com/article/30731204310/
- replace the download url in the /milvus/sdk/cmake/ThirdPartyPackages.cmake, line No.194
  `set(ZLIB_SOURCE_URL "https://github.com/madler/zlib/archive/v1.2.11.tar.gz")`
  change the url to gitee address that you import in the step 1
- import the https://github.com/milvus-io/grpc-milvus to Gitee
-  replace the download url in the /milvus/sdk/cmake/ThirdPartyPackages.cmake, line No.194
  `set(GRPC_SOURCE_URL "https://github.com/milvus-io/grpc-milvus/archive/master.zip")`
  change the url to gitee address that you import in the step 1
compile the sdk project

**Reply by NoError404** [reply]
**My internet is not good , so my reply  is slow. I have tried what u said.but it still encount some errors.** 
**I run milvus in docker**


My changes in the ThirdPartyPackages.cmake
if (DEFINED ENV{MILVUS_GRPC_URL})
    set(GRPC_SOURCE_URL "$ENV{MILVUS_GRPC_URL}")
else ()
    set(GRPC_SOURCE_URL
            "https://gitee.com/yhmo/grpc-milvus/repository/archive/master.zip")
endif ()

if (DEFINED ENV{MILVUS_ZLIB_URL})
    set(ZLIB_SOURCE_URL "$ENV{MILVUS_ZLIB_URL}")
else ()
    set(ZLIB_SOURCE_URL "https://gitee.com/yhmo/zlib/repository/archive/master.zip")
endif ()

**the error message shows:**

> 

[ 24%] No update step for 'grpc_ep'
[ 24%] No patch step for 'grpc_ep'
[ 26%] No configure step for 'grpc_ep'
[ 28%] Performing build step for 'grpc_ep'

> 

EPENDENCY ERROR
make[3]: *** [Makefile:993：stop] 错误 1
make[3]: *** 正在等待未完成的任务....

The target you are trying to run requires an OpenSSL implementation.
Your system doesn't have one, and either the third_party directory
doesn't have it, or your compiler can't build BoringSSL.

Please consult INSTALL to get more information.

If you need information about why these tests failed, run:

  make run_dep_checks

make[2]: *** [CMakeFiles/grpc_ep.dir/build.make:115：grpc_ep-prefix/src/grpc_ep-stamp/grpc_ep-build] 错误 2
make[1]: *** [CMakeFiles/Makefile2:285：CMakeFiles/grpc_ep.dir/all] 错误 2
make: *** [Makefile:130：all] 错误 2

**Reply by yhmo** [reply]
Did you see some messages like "Retry after x seconds"? 
Or "error: downloading 'https://gitee.com/yhmo/zlib/repository/archive/master.zip' failed"?

**Reply by fishpenguin** [reply]
Did you have `libcurl4-openssl-dev` installed in your system?

**Reply by NoError404** [reply]
**this is all its shows**
![image](https://user-images.githubusercontent.com/32030695/113390401-37902a00-93c4-11eb-887d-6481647add1f.png)
![image](https://user-images.githubusercontent.com/32030695/113390436-4a0a6380-93c4-11eb-9467-34e59ec14915.png)

**Is there something wrong with the installing of the grpc**

**Reply by yhmo** [reply]
Try:
`apt-get install pkg-config`

**Reply by NoError404** [reply]
i tried before , but didn't worked

**Comment by yhmo** [comment]
What is your cmake version?
`cmake --version`

Try compile a cmake with option "--system-curl":

```
wget https://github.com/Kitware/CMake/releases/download/v3.14.0/cmake-3.14.0.tar.gz
tar -zxv -f ./cmake-3.14.0.tar.gz
cd cmake-3.14.0
./bootstrap --system-curl
make
make install
```

**Reply by NoError404** [reply]
![image](https://user-images.githubusercontent.com/32030695/113401209-9ad68800-93d5-11eb-92aa-f89aaad8d681.png)

error still exits

**Reply by yhmo** [reply]
I also have no idea why it doesn't work. You can try search more information from google.

**Comment by NoError404** [comment]
before run  the ./build.sh  , i run sudo apt-get install libssl-dev .All worked.  Thank u for all of u!
refer this!
[https://stackoverflow.com/questions/16248775/cmake-not-able-to-find-openssl-library](url)

**Reply by yhmo** [reply]
Great!

In fact, in milvus main folder, there is a **ubuntu_build_deps.sh** to install dependencies for milvus, libssl-dev is installed by the sh file. But the C++ sdk readme doesn't mentioned this, I will add it to the readme.

**Reply by NoError404** [reply]
It must help!
